src/main/java/io/axoniq/axonserver/connector/ResultStreamPublisher.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2022. AxonIQ
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.axoniq.axonserver.connector;
+
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+/**
+ * {@link Publisher} implementation that allows to interact with a {@link ResultStream}.
+ * <p>
+ * When subscribed, it generates a new {@link ResultStream} through the provided supplier.
+ * The generated {@link ResultStream} provides the elements to be published.
+ *
+ * @author Sara Pellegrini
+ * @since 4.6
+ */
+public class ResultStreamPublisher<Message> implements Publisher<Message> {
+
+    private static final Logger logger = LoggerFactory.getLogger(ResultStreamPublisher.class);
+    private final Supplier<ResultStream<Message>> resultStreamSupplier;
+
+
+    public ResultStreamPublisher(Supplier<ResultStream<Message>> resultStreamSupplier) {
+        this.resultStreamSupplier = resultStreamSupplier;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super Message> s) {
+        ResultStreamSubscription subscription = new ResultStreamSubscription(s, resultStreamSupplier.get());
+        s.onSubscribe(subscription);
+        subscription.afterSubscribe();
+    }
+
+    private class ResultStreamSubscription implements Subscription {
+
+        private final Subscriber<? super Message> subscriber;
+        private final ResultStream<Message> resultStream;
+        private final AtomicLong requested = new AtomicLong(0);
+        private final AtomicBoolean signalGate = new AtomicBoolean(false);
+        private final AtomicBoolean cancelled = new AtomicBoolean(false);
+        private final AtomicBoolean subscribed = new AtomicBoolean(false);
+        private final AtomicBoolean completed = new AtomicBoolean(false);
+
+        private ResultStreamSubscription(Subscriber<? super Message> subscriber,
+                                         ResultStream<Message> resultStream) {
+            this.subscriber = subscriber;
+            this.resultStream = resultStream;
+            this.resultStream.onAvailable(this::signal);
```

Comment:
Having callbacks triggered to an object that hasn't been constructed completely yet is generally not recommended.
In a comparable situation, we have chosen to register this callback in the request method, with the following pseudo-code:
```public void request(long n) {
  if (requested.getAndUpdate(prev -> prev + n) == 0) {
      this.resultStream.onAvailable(this::signal);
  } 
  signal();
}```
The resultstream will only remember the last registered callback, so there is only very little overhead in this operation.

Reply:
what about moving that line into afterSubscribe method?

private void afterSubscribe() {
    subscribed.set(true);
    this.resultStream.onAvailable(this::signal);
    signal();
}

In this way, it is executed only once.